### PR TITLE
Use Config class instead of base class kwargs in structured config to not break early pydantic versions

### PR DESCRIPTION
**Pull request by schrockn**
### Summary & Motivation

We were using base class kwargs arguments in Pydantic in the structured config workstream. This breaks pydantic <1.8. Using `Config` class instead of kwargs.

### How I Tested These Changes

```
dagster-x86-3.10.3) elementl-13in-m1-mbppro-2021:script_to_assets_branching_io_manager schrockn$ pip list | grep pydantic
pydantic                            1.7.4
(dagster-x86-3.10.3) elementl-13in-m1-mbppro-2021:script_to_assets_branching_io_manager schrockn$ pytest /Users/schrockn/code/dagster/python_modules/dagster/dagster_tests/core_tests/structured_config_tests/test_basic_structured_config.py

This error occurred:
....
TypeError: MakeConfigCacheable.__init_subclass__() takes no keyword arguments

(dagster-x86-3.10.3) elementl-13in-m1-mbppro-2021:script_to_assets_branching_io_manager schrockn$ git co fix-pydantic
Switched to branch 'fix-pydantic'
(dagster-x86-3.10.3) elementl-13in-m1-mbppro-2021:script_to_assets_branching_io_manager schrockn$ !pyt
pytest /Users/schrockn/code/dagster/python_modules/dagster/dagster_tests/core_tests/structured_config_tests/test_basic_structured_config.py
WARNING:root:Unable to detect CI environment.  No test analytics will be sent.
================================================================================= test session starts ==================================================================================
platform darwin -- Python 3.10.3, pytest-7.0.1, pluggy-1.0.0
rootdir: /Users/schrockn/code/dagster, configfile: pyproject.toml
plugins: forked-1.4.0, structlog-0.5, ddtrace-1.6.1, buildkite-test-collector-0.1.3, dependency-0.5.1, requests-mock-1.10.0, snapshottest-0.6.0, xdist-2.1.0, mock-3.3.1, cov-2.10.1, respx-0.20.0, anyio-3.6.2, rerunfailures-10.0
collected 22 items

../../../python_modules/dagster/dagster_tests/core_tests/structured_config_tests/test_basic_structured_config.py ......................                                          [100%]

================================================================================== 22 passed in 0.59s ==================================================================================
```

### CODE REVIEW

**Comment by schrockn** [comment]
Current dependencies on/for this PR:
* master
  * **PR #11546** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11546" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/11546?utm_source=stack-comment).

File: python_modules/dagster/dagster/_config/structured_config.py [hunk 1]
```diff
@@ -63,7 +65,12 @@ class Config(MakeConfigCacheable):
     """
 
 
-class PermissiveConfig(Config, extra="allow"):
+class PermissiveConfig(Config):
+    # Pydantic config for this class
+    # Cannot use kwargs for base class as this is not support for pydnatic <1.8
+    class Config:
+        extra = "allow"
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_config/structured_config.py:72]: are you confident the subclassing here works the way it did with the kwargs approach

File: python_modules/dagster/dagster/_config/structured_config.py [hunk 2]
```diff
@@ -30,21 +30,23 @@ class cached_property:  # type: ignore[no-redef]
...
-    frozen=True,
-):
+class MakeConfigCacheable(BaseModel):
     """This class centralizes and implements all the chicanery we need in order
     to support caching decorators. If we decide this is a bad idea we can remove it
     all in one go.
     """
 
+    # Pydantic config for this class
+    # Cannot use kwargs for base class as this is not support for pydnatic <1.8
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_config/structured_config.py:40]: typo on pydantic

**Review by gibsondan — APPROVED** [review]
if you want to guard against this in the future you could make a tox suite ala definitions_tests_old_pendulum , not sure if warranted here or not https://github.com/dagster-io/dagster/blob/1abe25e78156c95a5423be164b4bd713454a7cf1/python_modules/dagster/tox.ini#L13

File: python_modules/dagster/dagster/_config/structured_config.py (hunk 1 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_config/structured_config.py:72]: Good call. Added unit test.

**Pull request merged**